purchase.py

```diff
@@ -23,71 +24,79 @@ def process(cls, purchases):
         super(Purchase, cls).process(purchases)
 
         if to_process:
-            self_companies = {x.party.id for x in Company.search([])}
+            companies = Company.search([])
+            party_by_companies = dict((c.party.id, c) for c in companies)
+            to_create_by_company = {company: [] for company in companies
+                                        if company.intercompany_user}
 
-            to_create = []
+            # create sales grouping by company (user and context from each company)
             for purchase in to_process:
-                if purchase.party.id not in self_companies:
+                company = party_by_companies.get(purchase.party.id)
+                if company and company.intercompany_user:
+                    to_create_by_company[company] += [purchase]
+
+            for company, purchases in to_create_by_company.items():
+                if not purchases:
                     continue
-                new_sale = purchase.create_intercompany_sale()
-                if new_sale:
-                    to_create.append(new_sale)
-            if to_create:
-                Sale.save(to_create)
+                with Transaction().set_user(company.intercompany_user.id):
+                    context = User.get_preferences(context_only=True)
+                    # sure set context has the company
+                    context['company'] = company.id
+                    context['companies'] = [company.id]
+                    context['_check_access'] = False
+
+                with Transaction().set_user(company.intercompany_user.id), \
+                    Transaction().set_context(context):
+                        to_create = []
+                        for purchase in purchases:
+                            new_sale = purchase.create_intercompany_sale()
+                            if new_sale:
+                                to_create.append(new_sale)
+                        if to_create:
+                            Sale.save(to_create)
 
     def create_intercompany_sale(self):
         pool = Pool()
         Party = pool.get('party.party')
         Sale = pool.get('sale.sale')
-        Company = pool.get('company.company')
 
-        company, = Company.search([('party', '=', self.party.id)], limit=1)
-        if not company.intercompany_user:
+        sale = Sale.search([
+            ('reference', '=', self.number)], limit=1)
```

Comment:
I understand that this is unchanged from existing code but the search must include the party. Otherwise there may be sales with the same reference but for other parties that would inhibit the creation of the new sale.

Reply:
Search sales by reference and party
7fb4e080c73b2bca650e24d7bab22bbcc902dd6d